Model: Fix some type errors

adityasharad opened the pull request:
## Description
This PR makes a start at cleaning up some TypeScript compile errors.

Add explicit array and number checks.
Add type annotations to the update functions based on the current/cumulative data types.

## Impacted Areas in the application
Type annotations and some runtime type checking when running the model. Should match the current expectations of the JS types being received.

## Testing
Unit tests and anywhere that runs the model.

adityasharad commented on src/algorithms/model.ts:
```diff
@@ -105,13 +107,43 @@ function advanceState(
+    const currentArray = pop[kind][compartment];
+    const deltaArray = tdot[kind][compartment];
+    const newArray = newPop[kind][compartment];
+    if (Array.isArray(currentArray) && Array.isArray(deltaArray) && Array.isArray(newArray)) {
```
Sadly doing this in a `forEach` prevents the TypeScript compiler from detecting that each value has array type.

nnoll commented:
This looks good to me - the types are as I imagined when I wrote the JS file :) @ivan-aksamentov and I discussed larger changes to the codebase (see PR #596) to solve this but this is ultimately simpler and starts to fix the hole in our types.

I'll let @ivan-aksamentov have final say but I vote we merge.

ivan-aksamentov commented:
@adityasharad @nnoll This is not what we want.

The first half of it adds runtime checks in the innermost loop of the algorithm, which makes me wonder how's performance on high values of `numbeStochasticRuns`. I'd expect a pretty significant hit. Just look at the amount of branches there. If @adityasharad can preform some profiling before and after that'd be great.

The second half is just two `as` type assertions (casts) and should be removed.

If we wanted to silence the checks we could just add the file to the compiler's exclude list.

adityasharad commented:
@ivan-aksamentov Good point. I have simplified the type checks somewhat and dropped the runtime checks, which should now give you valid type signatures on `update` and `updateAt` without incurring the additional runtime cost. Does that look reasonable?

> If we wanted to silence the checks we could just add the file to the compiler's exclude list.

For context, I'm submitting these changes because very soon I plan to contribute a simple CLI entry point (that invokes `run`), which we are already using in the GitHub unified UI to run your model. I'm using type correctness as a way of verifying that I've correctly interpreted the input/output schema of your model when calling it, and so would prefer to help clear up the existing type errors rather than exclude the file. Happy to keep receiving suggestions so I don't interfere with the workings of your model while I do that!
